README.md
# AWS MFA

![GitHub issues](https://img.shields.io/github/issues/quicken/aws-mfa)

This command-line tool automatically creates a temporary profile inside the credentials file that has been authenticated with MFA.

Several variations of this type of tool already exist in various programming languages with a mix of features.

I needed something that worked with node, was less opinionated and provided some flexibility missing in some of the alternatives.

Hope this tool makes your life easier in AWS Land.

Cheers

Marcel 😎👍

- #herdingbits
- Youtube: https://youtube.com/herdingbits
- Blog: https://www.herdingbits.com

![Your Repository's Stats](https://github-readme-stats.vercel.app/api?username=quicken&show_icons=true&hide=commits,prs,issues,contribs)

## Prerequisites

Requires node version 16+

Ensure that you have already configured a valid AWS Credentials file.

### See the official AWS Documentation for defining the credentials file:

https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html

### See the official AWS Documentation for using MFA:

https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-role.html#cli-configure-role-mfa

## Installation

https://www.npmjs.com/package/@herdingbits/awsmfa

For npm users:

```
npm i -g @herdingbits/awsmfa
```

For yarn users:

```
yarn global add @herdingbits/awsmfa
```

If you would like to try without installing it first:

```
npx @herdingbits/awsmfa
```

## Usage

1. Run the tool to automatically create a "mfa" profile inside your credentials file using your MFA session.
2. Use the created "mfa" profile anywhere you can specify an AWS profile.

This is especially useful inside the ~/.aws/config file as a "source_profile" for assuming roles securely.

However, it is also a good fit for working securely with the AWS-CLI, Serverless Offline as well as the AWS SDKs.

For basic help run:

```bash
awsmfa --help
```

### Basic Interactive Usage

For a basic setup, add your serial number to the profile that is linked to your MFA device to your **~/.aws/credentials** file.

**~/.aws/credentials**

```ini
[default]
aws_access_key_id=MYACCESSKEY000000000
aws_secret_access_key=thisisthesecretaccesskey01234567890ABCDE
mfa_serial=arn:aws:iam::012345678901:mfa/myuser
```

Run the command and when prompted, enter the MFA token code shown on your device.

```bash
awsmfa
```

The tool will not display any messages if successful. You can now use the automatically generated or update "mfa" profile to make calls to the AWS APIs.
For example, using the AWS CLI:

```bash
aws s3 ls --profile mfa
```

### Advanced Usage

_WARNING: This advanced use case will overwrite your default profile credentials._

In this scenario, you configure a "login" profile with your AWS credentials, **awsmfa** will then update your "default" profile with the temporary credentials.

**~/.aws/credentials**

```ini
[login]
aws_access_key_id=MYACCESSKEY000000000
aws_secret_access_key=thisisthesecretaccesskey01234567890ABCDE
mfa_serial=arn:aws:iam::012345678901:mfa/myuser
```

Run the command and when prompted, enter the MFA token code shown on your device.

```bash
# WARNING: This advanced use case will overwrite your default profile credentials.
awsmfa -p login --mfa-profile default
```

You can now just run any AWS command and it will use the default profile which is set to use your MFA session.

For example, using the AWS CLI:

```bash
aws s3 ls
```

Alternatively, if you use profiles to "assume roles". Configure your "profiles" in the **~/.aws/config** file as follows for easy MFA usage.

**~/.aws/config**

```ini
[profile devops]
region = ap-southeast-2
role_arn = arn:aws:iam::012345678901:role/devops-crew
source_profile = mfa
```

You can now authenticate with "awsmfa" and seamlessly use your "devops" profile to assume roles.

### Non-Interactive

The tool can also be run non-interactive for further automation or scripting.

In this example we enter the MFA code programmatically, authenticate with the "foo" profile and write our MFA credentials to the profile "coolio"

```bash
awsmfa -t 012345 -p foo --mfa-profile coolio
```

## Command-line Arguments

| Argument               | Description                                                                                                                                                                                                |
| ---------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| -d, --duration-seconds | The duration in seconds, that the credentials should remain valid. Valid ranges are between 900 seconds (15 minutes) to 129,600 seconds (36 hours), with 43,200 seconds (12 hours) as the default.         |
| -h, --help             | Show help.                                                                                                                                                                                                 |
| -p, --profile          | Use the specific profile from the credentials file for getting the session. Note: This works differently to the AWS CLI in that profiles defined in the ".aws/config" file are ignored.                    |
| -s, --serial-number    | The identification number of the MFA device that is associated with the IAM user requiring MFA. The serial number can also be set for the profile in the credentials file using the "mfa_serial" property. |
| -t, --tokenCode        | The value provided by the MFA device. If this argument is omitted the command will pause and prompt the user for the token code.                                                                           |
| --mfa-profile          | The profile that temporary credentials are written to. Defaults to "mfa".                                                                                                                                  |
| --version              | Display version information.                                                                                                                                                                               |

## Change Logs

https://github.com/quicken/aws-mfa/blob/master/CHANGELOG.md

Please report any errors or make feature requests on GitHub.

https://github.com/quicken/aws-mfa/issues


CHANGELOG.md
# Change Log

All notable changes to the "awsmfa" command will be documented in this file.

Please submit any issues you find to Git-Hub or better yet submit a pull request.

https://github.com/quicken/aws-mfa

### 1.0.3

- Fix a bug that prevented entering token codes starting with zero.

### 1.0.1

- The prompt is now more concise.
  Allow specifying the "default" profile as an MFA target profile.


__tests__/demo.test.ts
import { expect, test } from "@jest/globals";
import "dotenv/config";

test("simple_math", () => {
	const answer = 1 + 2;
	expect(answer).toBe(3);
});


__tests__/ConfigTools.test.ts
import { expect, test } from "@jest/globals";
import "dotenv/config";
import { ConfigTool, CommandConfiguration, CredentialsEntry } from "../src/ConfigTool";
import { ValidationException } from "../src/ValidationException";

test("validate durationSeconds", () => {
    const credentialsEntry = getCredentialsEntry();
    const config = getDefaultConfig();

    /* Below Bounds */
    try {
        config.durationSeconds = 899;
        ConfigTool.validate(config, credentialsEntry);

        throw new Error("no error thrown when durationSeconds is below bounds.");
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
    }

    /* Above Bounds */
    try {
        config.durationSeconds = 129601;
        ConfigTool.validate(config, credentialsEntry);

        throw new Error("no error thrown when durationSeconds is above bounds.");
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
    }

    /* On lower boundary */
    try {
        config.durationSeconds = 900;
        ConfigTool.validate(config, credentialsEntry);
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
        throw new Error("error thrown on lower boundary.");
    }

    /* In range boundary */
    try {
        config.durationSeconds = 43200;
        ConfigTool.validate(config, credentialsEntry);
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
        throw new Error("error thrown in range.");
    }

    /* On upper boundary */
    try {
        config.durationSeconds = 129600;
        ConfigTool.validate(config, credentialsEntry);
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
        throw new Error("error thrown on upper boundary.");
    }
});

test("validate tokenCode", () => {
    const credentialsEntry = getCredentialsEntry();
    const config = getDefaultConfig();

    /* token to short */
    try {
        config.tokenCode = "01234";
        ConfigTool.validate(config, credentialsEntry);

        throw new Error("no error thrown when token is too short.");
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
    }

    /* token too long */
    try {
        config.tokenCode = "0123456";
        ConfigTool.validate(config, credentialsEntry);

        throw new Error("no error thrown when token is too long");
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
    }

    /* token undefined. */
    try {
        config.tokenCode = undefined;
        ConfigTool.validate(config, credentialsEntry);
        throw new Error("no error thrown when token is undefined.");
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
    }

    /* valid token */
    try {
        config.tokenCode = "012345";
        ConfigTool.validate(config, credentialsEntry);
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
        throw new Error("error thrown on valid token.");
    }
});

test("validate prevent overwriting profile", () => {
    const credentialsEntry = getCredentialsEntry();
    const config = getDefaultConfig();

    /* If profile and mfa profile are equal this could oeverwrite credentials.
	this should be prevented. */
    try {
        config.profile = "testing";
        config.mfaProfile = "testing";
        ConfigTool.validate(config, credentialsEntry);

        throw new Error("no error preventing profile and mfa profile from being equal.");
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
    }
});

test("validate credentials not found", () => {
    const config = getDefaultConfig();

    try {
        ConfigTool.validate(config, undefined);

        throw new Error("no error when credentials are undefined.");
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
    }
});

test("validate mfa_serial", () => {
    /* Missing mfa serial */
    try {
        const config = getDefaultConfig();
        config.serialNumber = undefined;
        const credentialsEntry = getCredentialsEntry();
        credentialsEntry.mfa_serial = undefined;
        ConfigTool.validate(config, credentialsEntry);

        throw new Error("no error when serial number is undefined in config and credentials.");
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
    }

    /* Missing mfa serial in credentials*/
    try {
        const config = getDefaultConfig();
        config.serialNumber = "arn:aws:iam::012345678901:mfa/myuser";
        const credentialsEntry = getCredentialsEntry();
        credentialsEntry.mfa_serial = undefined;
        ConfigTool.validate(config, credentialsEntry);
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
        throw new Error("invalid error when serial number is defined in config and not in credentials.");
    }

    /* Missing mfa serial in config*/
    try {
        const config = getDefaultConfig();
        config.serialNumber = undefined;
        const credentialsEntry = getCredentialsEntry();
        credentialsEntry.mfa_serial = "arn:aws:iam::012345678901:mfa/myuser";

        ConfigTool.validate(config, credentialsEntry);
    } catch (e) {
        expect(e instanceof ValidationException).toBe(true);
        throw new Error("invalid error when serial number is defined in credentials and not in config.");
    }
});

function getDefaultConfig(): CommandConfiguration {
    return {
        tokenCode: "012345",
        profile: "default",
        durationSeconds: 43200,
        serialNumber: "arn:aws:iam::012345678901:mfa/myuser",
        mfaProfile: "mfa"
    };
}

function getCredentialsEntry(): CredentialsEntry {
    return {
        aws_access_key_id: "MYACCESSKEY000000000",
        aws_secret_access_key: "thisisthesecretaccesskey01234567890ABCDE"
    };
}


src/index.ts
#!/usr/bin/env node

import fs from "fs";
import ini from "ini";
import os from "os";
import { Credentials } from "@aws-sdk/types";
import { STS, GetSessionTokenCommandInput, STSServiceException } from "@aws-sdk/client-sts";
import { ConfigTool, CredentialsEntry } from "./ConfigTool";
import { ValidationException } from "./ValidationException";

(async function () {
    try {
        const credentialsFile = os.homedir() + "/.aws/credentials";
        const store = ini.parse(fs.readFileSync(credentialsFile, "utf-8"));

        /* #####################################*/
        /* Get configuration from args and/or by prompting the user. */
        const config = await ConfigTool.readArgs();
        if (!config.tokenCode) {
            config.tokenCode = (await ConfigTool.promptForToken(config.profile)).toString();
        }

        const credentialsEntry = store[config.profile] as CredentialsEntry;
        config.serialNumber = config.serialNumber ? config.serialNumber : credentialsEntry.mfa_serial;

        ConfigTool.validate(config, credentialsEntry);

        /* #####################################*/
        /* Fetch STS Token */
        const rootCredentials: Credentials = {
            accessKeyId: credentialsEntry.aws_access_key_id,
            secretAccessKey: credentialsEntry.aws_secret_access_key
        };

        const sts = new STS({
            credentials: rootCredentials
        });

        const getSessionTokenCommandInput: GetSessionTokenCommandInput = {
            SerialNumber: config.serialNumber,
            TokenCode: config.tokenCode + "",
            DurationSeconds: config.durationSeconds
        };

        const getSessionTokenCommandOutput = await sts.getSessionToken(getSessionTokenCommandInput);
        if (getSessionTokenCommandOutput.Credentials) {
            /* #####################################*/
            /* Write temporary credentials back to credentials file. */
            const mfaCredentials = getSessionTokenCommandOutput.Credentials;
            if (!store[config.mfaProfile]) {
                store[config.mfaProfile] = {};
            }
            store[config.mfaProfile].aws_access_key_id = mfaCredentials.AccessKeyId;
            store[config.mfaProfile].aws_secret_access_key = mfaCredentials.SecretAccessKey;
            store[config.mfaProfile].aws_session_token = mfaCredentials.SessionToken;
            fs.writeFileSync(credentialsFile, ini.stringify(store, {}));
        }
    } catch (e) {
        if (e instanceof STSServiceException) {
            console.log("\x1b[31m%s\x1b[0m", e.message);
        } else if (e instanceof ValidationException) {
            console.log("\x1b[31m%s\x1b[0m", e.message);
        } else {
            console.log(e);
        }

        process.exit(1);
    }
})();


src/ConfigTool.ts
import yargs from "yargs";
import prompt, { Schema } from "prompt";
import colors from "colors";
import { ValidationException } from "./ValidationException";

/**
 * Defines the configuration that is used when executing this command.
 */
export type CommandConfiguration = {
    tokenCode?: string;
    profile: string;
    durationSeconds: number;
    serialNumber?: string;
    mfaProfile: string;
};

/**
 * Defines values that are defined inside the profile section of the AWS credentials file.
 * NOTE: The typing currently only includes values needed by this tool and not the full set of
 * possible properties that could exist inside a AWS Credentials file.
 */
export type CredentialsEntry = {
    aws_access_key_id: string;
    aws_secret_access_key: string;
    mfa_serial?: string;
};

/**
 * Helper class for getting the commands configuration setting from arguments and/or from prompting the user.
 */
export class ConfigTool {
    /**
     * Read in command line arguments.
     * @returns Command Configuration
     */
    static readArgs = async (): Promise<CommandConfiguration> => {
        const options = await yargs
            .usage("Usage: -t <token-code>")
            .option("s", {
                alias: "serial-number",
                describe:
                    "The identification number of the MFA device that is associated with the IAM user requiring MFA.",
                type: "string",
                demandOption: false
            })
            .option("d", {
                alias: "duration-seconds",
                describe:
                    "The duration in seconds, that the credentials should remain valid. Valid ranges are between 900 seconds (15 minutes) to 129,600 seconds (36 hours), with 43,200 seconds (12 hours) as the default.",
                type: "number",
                demandOption: false
            })
            .option("p", {
                alias: "profile",
                describe:
                    "Use a specific profile from the credentials file. Note: This works differently to the AWS CLI in that profiles defined in the .aws/config file are ignored.",
                type: "string",
                demandOption: false
            })
            .option("t", {
                alias: "token-code",
                describe: "The value provided by the MFA device.",
                type: "string",
                demandOption: false
            })
            .option("mfa-profile", {
                describe: 'The profile that temporary credentials are written to. Defaults to "mfa".',
                type: "string",
                demandOption: false
            }).argv;

        const {
            t: tokenCode,
            p: profile = "default",
            d: durationSeconds = 43200,
            s: serialNumber,
            "mfa-profile": mfaProfile = "mfa"
        } = options;

        return {
            tokenCode: tokenCode,
            profile: profile,
            durationSeconds: durationSeconds,
            serialNumber: serialNumber,
            mfaProfile: mfaProfile
        };
    };

    /**
     * Prompt the user for the currently displayed MFA token.
     *
     * @param profile The profile for which the user is prompted.
     * @returns The token-code shown on the users device. (as entered by the user.)
     */
    static promptForToken = async (profile: string) => {
        const schema: Schema = {
            properties: {
                tokenCode: {
                    description: colors.green(`MFA code for "${profile}"`),
                    pattern: /^\d{6}$/,
                    type: "string",
                    message: colors.red("The token must be exactly 6 digits."),
                    required: true
                }
            }
        };
        prompt.start();
        const { tokenCode } = await prompt.get(schema);
        return tokenCode;
    };

    /**
     * Validates a command configuration. If validation fails a ValidationException error is thrown.
     *
     * @param config The command configuration that should be validated.
     * @param credential The credentials section that is being used as the source for RootCredentials.
     */
    static validate = (config: CommandConfiguration, credential: CredentialsEntry | undefined) => {
        if (config.durationSeconds < 900 || config.durationSeconds > 129600 || isNaN(config.durationSeconds)) {
            throw new ValidationException(
                "Invalid duration. Duration must be a value between 900 (15 minutes) and 129600 (36 hours)."
            );
        }

        if (!config.tokenCode || config.tokenCode.length != 6) {
            throw new ValidationException(
                "The token must be numeric and exactly 6 digits. Token is the token code that is displayed on your MFA device."
            );
        }

        if (config.profile === config.mfaProfile) {
            throw new ValidationException(
                "Profile and MFA Profile can not be the same value to prevent overwriting your credentials."
            );
        }

        if (!credential) {
            throw new ValidationException(`Unknown profile ${config.profile}. Does this profile exist?`);
        }

        if (!credential.mfa_serial && !config.serialNumber) {
            throw new ValidationException(
                'Unknown mfa serial number. Either define the parameter "mfa_serial" as a part of the profile within the credentials file or pass the mfa serial number with the -s argument.'
            );
        }
    };
}
